runtime/util/hshelp.c

```diff
@@ -492,8 +669,55 @@ areDoubleSlotConstantRefsIdentical(J9ROMConstantPoolItem * romCP1, U_32 index1,
 	return (ref1->slot1 == ref2->slot1) && (ref1->slot2 == ref2->slot2);
 }
 
+/**
+ * Compares two methods bytecode by bytecode to determine equivalence.
+ * 
+ * @param method1 first method to compare
+ * @param romClass1 ROM class associated with method1
+ * @param method2 second method to compare
+ * @param romClass2 ROM class associated with method2
+ * @return true if equivalent, false otherwise
+ */
 UDATA
 areMethodsEquivalent(J9ROMMethod * method1, J9ROMClass * romClass1, J9ROMMethod * method2, J9ROMClass * romClass2)
+{
+	return areMethodsEquivalentSub(method1, romClass1, NULL, method2, romClass2, NULL);
+}
+
+/**
+ * Compares two methods bytecode by bytecode to determine equivalence.
+ * For invokedynamic if the referenced callsite was resolved in method1
+ * the resolved MethodHandle will be copied to the RAM class callsite table
+ * in method2.
+ *
+ * @param method1 first method to compare
+ * param ramClass1 RAM class associated with method1
+ * @param method2 second method to compare
+ * @param ramClass2 RAM class associated with method2
+ * @return true if equivalent, false otherwise
+ */
+static UDATA
+areMethodsEquivalentPropagateCallSites(J9ROMMethod * method1, J9Class *ramClass1, J9ROMMethod * method2, J9Class *ramClass2) 
+{
+	return areMethodsEquivalentSub(method1, ramClass1->romClass, ramClass1, method2, ramClass2->romClass, ramClass2);
+}
+
+/**
+ * Compares two methods bytecode by bytecode to determine equivalence. If invoke
+ * dynamic callsites are determined to be equivalent and RAM classes are not null, resolved
+ * callsites will be propagated from ramClass1 to ramClass2. It is assumed that callsites will
+ * be propagated from 1 (original) to 2 (new).
+ * 
+ * @param method1 first method to compare
+ * @param romClass1 ROM class associated with method1
+ * @param ramClass1 RAM class associated with method1, must be provided to propagate callsites
+ * @param method2 second method to compare
+ * @param romClass2 ROM class associated with method2
+ * @param ramClass2 RAM class associated with method2, must be provided to propagate callsites
+ * @return true if equivalent, false otherwise
+ */
+static UDATA
+areMethodsEquivalentSub(J9ROMMethod * method1, J9ROMClass * romClass1, J9Class * ramClass1, J9ROMMethod * method2, J9ROMClass * romClass2, J9Class * ramClass2)
```

Comment:
Do we need the ROM and RAM parameters (i.e. can ROM be fetch from RAM, or has there already been some field flipping which would make that invalid)?

Reply:
The main reason for it is having a NULL RAM class input indicates that callsites don't need to be propagated when `areMethodsEquivalent` is called. `areMethodsEquivalentSub` is only supposed to be called by either `areMethodsEquivalent` or `areMethodsEquivalentPropagateCallSites`.